File: pepcoding/scala/L2/ArrayStrings/sortByParity.scala
object sortByParity{

    def sortArrayByParity(nums: Array[Int]): Array[Int] = {
        var even = 0
        var odd = nums.length - 1

        while (even < odd)
            if (nums(even) % 2 == 0)
                even = even + 1
            else {
                val temp = nums(odd)
                nums(odd) = nums(even)
                nums(even) = temp
                odd = odd - 1
            }
        nums
    }

    def main(args: Array[String]): Unit = {
        print(sortArrayByParity(Array(3,1,2,4)))
    }
}

File: pepcoding/java/L2/ArrayStrings/sortByParity.java
import java.util.*;

public class sortByParity {

    // ~~~~~~~~~~~~~~~~~~User Section~~~~~~~~~~~~~~~~~
    
    public static void sortArrayByParity(int[] nums) {
        // write your code here
        int i=0; // first unsolved
        int j=0; // first odd
        int n = nums.length;
        
        while(i < n){
            if(nums[i] % 2 == 0){
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
                i++;
                j++;
            }
            else{
                i++;
            }
        }
    }

    //// ~~~~~~~~~~~~~~~~~~Input Management~~~~~~~~~~~~~~~~~
    public static void main(String[] args) {
        int[] arr = {3,1,2,4};
        sortArrayByParity(arr);
        for(int i = 0; i < n; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
